src/compiler/utilities.ts

```diff
@@ -1792,6 +1792,11 @@ namespace ts {
         return containerKind === SyntaxKind.InterfaceDeclaration || containerKind === SyntaxKind.TypeLiteral;
     }
 
+    export function isFirstIdentifierOfImplementsClause(node: Node) {
+        return node.parent?.parent?.parent?.kind === SyntaxKind.HeritageClause

```

Comment:
why does it have to be the *first* identifier? Why can't it be anywhere?

(That long chain of `?.`s gives me the shivers.)

Reply:
Because this is a hot path and any other location in the node will be covered by other criteria of `isValidTypeOnlyAliasUseSite`, so it doesn’t make sense to check more than necessary.

`node` could be an identifier at the top level of a SourceFile, so I think that’d be SourceFile > ExpressionStatement > Identifier, so I guess only the last `?.` is needed.

Reply:
Also I don't think it checks the scenario of type ? ` types.SomeInterfaceContainingNamespace.Component` since its parent is going to be one more level up than just `types.Component` ?

Reply:
> Because this is a hot path and any other location in the node will be covered by other criteria of isValidTypeOnlyAliasUseSite,

double checking that this is true

Reply:
@sheetalkamat this only gets called on identifiers from `resolveName`, which will only be the left-most identifier. The other parts of entity names get checked elsewhere.